[Comments Query Loop] Show placeholder comments on site editor

cbravobernal opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
This is a quick approach to #37143. We are using the default placeholders' fallback of all components of a comment template.
A better solution may be to update all fallbacks to not depend on getting an empty comment object, feel free to comment or discuss if we should work on a better approach.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
- Go to site editor.
- Add a Comment Query Loop.
- Check that we are showing the default placeholders.
- Go to a post or a page.
- Add a Comment Query Loop.
- Check that comments loaded are real ones.

## Screenshots <!-- if applicable -->
Screenshots were done on site editor.
Before: 
![commentsBeforePR](https://user-images.githubusercontent.com/37012961/150137100-f0a4009e-4696-4c35-80df-364392c3c3c8.png)

After:

![Screenshot 2022-01-27 at 15 06 00](https://user-images.githubusercontent.com/37012961/151374813-15084512-f96a-4e7e-81ec-2d67aa80b001.png)



## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

New possible feature discussed on #37143

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->

Commit:
Add placeholder for comments block if there is no context id

Commit:
Update comment information of having empty object

gziolo commented:
As simple as that? 😄 I'll have a closer look tomorrow.

cbravobernal commented:
> As simple as that? 😄 I'll have a closer look tomorrow.

It seems so, but I don't know if is a good approach or we usually do setting empty objects as default in the project  😅

ntsekouras commented:
Hm.., interesting! It would require a bit more changes though, for example [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/comment-template/edit.js#L99) as it throws a React warning right now.

cbravobernal commented:
> Hm.., interesting! It would require a bit more changes though, for example [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/comment-template/edit.js#L99) as it throws a React warning right now.

Using null as id would remove the warning.

Anyway, the "options" are not working, and of course, if you set more than 1 item per page. It crashes (not very badly, though).

I will iterate it more.

Commit:
Prevent no key on array errors and multiple comments per page overload

cbravobernal commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -97,7 +102,7 @@ const CommentsList = ( {
 			comments.map( ( comment ) => (
 				<BlockContextProvider
-					key={ comment.commentId }
+					key={ comment.commentId || uniqueId( 'comment-default' ) }
```
This way, if we have a null id (required for loading placeholders and not calling REST API for comments), we don't have an error or warning regarding the unique key inside a list. Not sure if it is the best approach.

cbravobernal replied:
It seems that this approach is preventing the block to be selected.

Edited: it was a lack of a parent `div` with the `blockprops`, now it works fine.

gziolo replied:
I think we need to apply a fallback key as part of `comment` object to ensure those keys don't change for the individual items. Otherwise, we will have unnecessary rerenders happening too often.

ockham replied:
Can we use the `comments.map()`'s `index` as a fallback?

```suggestion
					key={ comment.commentId || index }
```

It's only considered an antipattern if individual items can be removed or moved around, which shouldn't be the case for placeholders IIUC.

cbravobernal commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -122,9 +127,20 @@ export default function CommentTemplateEdit( {
+			// to inner blocks to render the default placeholders.
+			if ( ! postId ) {
+				// We set a limit in order not to overload the editor of empty comments.
+				const defaultCommentsToShow = perPage <= 3 ? perPage : 1;
```
Is there any recommended number of comments placeholders to show at maximum?
Ping design @kjellr @jameskoster @karmatosed  😅

jameskoster replied:
Good question. My initial thought is to check `thread_comments_depth` and display enough placeholders to demonstrate maximum nesting.

michalczaplinski replied:
`thread_comments_depth` can go up to 10 though so that might be overkill 😅 

I've settled on a somewhat arbitrary maximum nesting of 3 for the placeholders. I think that should make the threading clear without overwhelming the UI.

jameskoster replied:
I don't know, I think that can be important. For example if you want to support a depth of 10, you may need to adjust the margins in order for that level of nesting to work. If you don't see a preview then you wouldn't be aware of such an issue.

Commit:
Add placeholder for comment avatar

michalczaplinski commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -125,6 +130,18 @@ export default function CommentTemplateEdit( {
+				// We set a limit in order not to overload the editor of empty comments.
+				const defaultCommentsToShow = perPage <= 3 ? perPage : 1;
+				return {
+					rawComments: Array( defaultCommentsToShow ).fill( {
+						id: null,
+					} ),
```
I was quite sure that this would fail if `defaultCommentsToShow` were larger than 1, but I was wrong. 

My concern was that [`convertToTree`](https://github.com/WordPress/gutenberg/blob/1f92999896beb98a572f46722e35b31b1de8d547/packages/block-library/src/comment-template/util.js#L30) would fail in that case, but it handles it just fine 👍

cbravobernal replied:
Let's see how will work with nested default comments (as it searches for a parent id) 😅

michalczaplinski commented:
Just gave it a spin. Looks great already! 👍 

https://user-images.githubusercontent.com/5417266/151464413-912090cf-f1f3-49ce-8478-50c4d99e36a6.mov

I assume you'll still have to add the `thread_comments_depth` to the REST API and read it in the editor like mentioned in https://github.com/WordPress/gutenberg/pull/38072#discussion_r792805208 but other than that I think it's almost ready to merge 🙂

jameskoster commented:
Not sure if this is something to include in this PR, but thinking about it a little more, we can probably use `avatar_default` for the Comment Author Avatar placeholder appearance. 

Then as a follow up, perhaps we can also make that option available in the block toolbar or inspector. Edit: Opened #38321 for this :)

Commit:
Add threaded options

michalczaplinski committed:
Comment Template: Add support for threaded comments in comment placeholders

michalczaplinski commented:
I've added support for threaded comments in the placeholders.

https://user-images.githubusercontent.com/5417266/151640092-cdc37bbb-3d07-4e4b-8e75-59db74c235ec.mp4

You might notice that there is some flicker and the Comment Template is rerendering needlessly. I ll work on this in the followup to https://github.com/WordPress/gutenberg/issues/37154

I have some reservations as to whether this https://github.com/WordPress/gutenberg/blob/fc62d9c6bf2389c5f17bdcef795ca70059d7aab9/packages/block-library/src/comment-template/edit.js#L162-L171 
is a good idea, but I couldn't think of a better way to handle this for now.

cbravobernal commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -119,12 +124,52 @@ export default function CommentTemplateEdit( {
+				const defaultCommentsToShow =
+					perPage <= commentsDepth ? perPage : commentsDepth;
+
+				// This is a little hacky. We get a comment structure like:
+				// [
+				//   {id: -1, parent: 0},
+				//   {id: -2, parent: -1},
+				//   {id: -3, parent: -2}
+				//  ]
+				// The idea is that comments with a negative ID will not appear in a
+				// "real" system.
```
🤯  😆

gziolo replied:
> I have some reservations as to whether this is a good idea, but I couldn't think of a better way to handle this for now.

It would be nice to extract all the logic that generates fake data and cover it with unit tests so it's easier to update if we ever have to. We already have `utils.js` file and some related tests.

I guess it's fine as long as it doesn't lead to some unexpected cases. One thing I'm still unsure about is the code check that happens later:

https://github.com/WordPress/gutenberg/blob/fc62d9c6bf2389c5f17bdcef795ca70059d7aab9/packages/block-library/src/comment-template/edit.js#L217-L219

When there is no `postId` set then this message will never be shown. In the case when there is `postId` set and there are no comments then it will still render this message. That's something to think about. Maybe we should never show this state in the UI 🤔

cbravobernal replied:
Another issue of this approach is that we try to fetch those negative id comments on the Rest API, having a 404 per placeholder.

![Screenshot 2022-01-31 at 11 28 01](https://user-images.githubusercontent.com/37012961/151777624-61428a60-311b-48ca-9361-772cfca43af2.png)

Also, it would be cool to tell the user that the threaded comments are replies, not everyone may know this when starting a FSE site.

gziolo replied:
This is something we definitely want to avoid. It would flood server logs 😅

gziolo commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -119,12 +124,52 @@ export default function CommentTemplateEdit( {
+			if ( ! postId ) {
+				// If displaying threaded comments is disabled, we only show one comment
+				if ( ! threadComments ) {
+					return { rawComments: [ { id: null } ] };
```
Should the `id` be different here `0` or `-1`? @michalczaplinski should know better.

michalczaplinski replied:
It should probably not be `0` because that can be a valid `id`, right? However, I think it does not matter probably, because we should find a way not to to make the calls to the REST API to fetch the "fake" comments at all like you and Carlos have [found out already](https://github.com/WordPress/gutenberg/pull/38072#discussion_r795535875) 😄

gziolo commented:
It looks like everything works as expected, so I don't see any reason why you shouldn't go with that initial approach. The concept of fake data is novel in the block editor, but it's hard to tell at the moment about the drawbacks of this proposal. I guess the concept of nested block template repeater is complex enough on its own 😄

cbravobernal commented:
> Not sure if this is something to include in this PR, but thinking about it a little more, we can probably use `avatar_default` for the Comment Author Avatar placeholder appearance.
> 
> Then as a follow up, perhaps we can also make that option available in the block toolbar or inspector. Edit: Opened #38321 for this :)

This may need some adjustments to the REST API in order to show those defaults. As far as I've researched, I only found functions to get the default avatars by user (not valid for this case).

We can do a follow-up PR after finishing this. So this one does not get much bigger ( we still have to improve the placeholder experience )

Commit:
Allow placeholder to work with border radius, remove background color from avatar

cbravobernal commented:
I updated the code in order for the border to work with the placeholder. Maybe is a little overkill having to unserialize (and with that using a default avatar image would be a better option). But [I took the work of @carolinan in post author ](https://github.com/WordPress/gutenberg/pull/38265/files), so at least we have the same structure in both blocks.

https://user-images.githubusercontent.com/37012961/151818447-6e9a63bf-0920-442a-8ca4-cfc0be1166e4.mov

Commit:
Use default avatars for placeholders

michalczaplinski commented:
I'm gonna take a look to see if we can avoid making the REST calls for comments that we know don't exist like mentioned in https://github.com/WordPress/gutenberg/pull/38072#discussion_r795535875. I don't have a good idea how to do it so far.

michalczaplinski commented:
> I'm gonna take a look to see if we can avoid making the REST calls for comments that we know don't exist like mentioned in [#38072 (comment)](https://github.com/WordPress/gutenberg/pull/38072#discussion_r795535875). I don't have a good idea how to do it so far.

OK, I took a look and I see 2 ways that we can go about fixing this issue:

1. Refactor the blocks that are inner blocks of the Comment Template (Comment Author, Comment Avatar, Comment Date, etc.) so that they don't make any call to `getEntityRecord()` if the `commentId` is negative. I think this is not the best idea because it relies too much on this undocumented convention of having negative comment ids in too many places.

2. Remove the code that I added which creates the comments with negative `id`s. Then define explicit "placeholder" state in the Comment Template block that will already have the correct shape like: 
   ```json
   {
     "commentId": null,
     "children": [
       { "commentId": null, "children": [{ "commentId": null, "children": [] }] },
       { "commentId": null, "children": [] }
     ]
   }
   ```

   The main (and really the only one) difference between this "placeholder" state and real data is is that all the `id`s are `null`.
   
   I ll add a commit with those changes.
   
---
   
On an unrelated note just a heads up that this branch is going to have some merge conflicts with changes done by @DAreRodz in https://github.com/WordPress/gutenberg/pull/38187  🙂

michalczaplinski committed:
Comment Template: Add a rough version of nested placeholder comments.

This implementation improves the previous one in that it does not call
the REST API with non-existent comment IDs

michalczaplinski commented:
Here we go: ca1c6973e6d07703774980195405910ce8b51

Works fine - the REST API is not being called with fake ids anymore. The implementation needs a little cleaning up though but I didn't want to invest time right now because this code is probably still going to change a bit because of the #38187 that David is still working on.

gziolo commented:
It looks like tested block fixtures need to be regenerated:

<img width="770" alt="Screenshot 2022-02-02 at 07 15 12" src="https://user-images.githubusercontent.com/699132/152102496-8d52fe81-1e48-417b-acea-c8de7f7de9e4.png">

More in https://github.com/WordPress/gutenberg/runs/5030469389?check_suite_focus=true.

It makes me wonder if the changes to the supports won't cause block invalidation. It needs to be tested. It might be not that bad at this stage, but we might need a temporary deprecation if any theme in the directory is using those blocks.

carolinan commented:
I could not find any themes in the directory that uses the new set of comments blocks.
I think it would be helpful to have a decision tree to help us know when to do deprecations. And also instructions for how to test them.

Commit:
Allow background color again to avoid compatibility errors

cbravobernal commented:
> It looks like tested block fixtures need to be regenerated:

I added again the background color so no deprecations are needed 😄 

I think the PR is ready to land. I will now work on an Avatar block (working wherever we put them).

jameskoster commented:
I'm seeing settings for blocks inside Comment Template repeated three times 😅
<img width="1612" alt="Screenshot 2022-02-07 at 10 53 28" src="https://user-images.githubusercontent.com/846565/152775021-ecc35b10-c722-4b55-8991-886abe1dfe2a.png">

cbravobernal commented:
> I'm seeing settings for blocks inside Comment Template repeated three times 😅 <img alt="Screenshot 2022-02-07 at 10 53 28" width="1612" src="https://user-images.githubusercontent.com/846565/152775021-ecc35b10-c722-4b55-8991-886abe1dfe2a.png">

That is already happening before the PR 😓  I [added an issue](https://github.com/WordPress/gutenberg/issues/38080), but it also was mentioned in [ this comment](https://github.com/WordPress/gutenberg/issues/37154#issuecomment-998728016)

I will work on that as soon as possible.

cbravobernal commented:
> It looks like tested block fixtures need to be regenerated:
> 
> <img alt="Screenshot 2022-02-02 at 07 15 12" width="770" src="https://user-images.githubusercontent.com/699132/152102496-8d52fe81-1e48-417b-acea-c8de7f7de9e4.png">
> 
> More in https://github.com/WordPress/gutenberg/runs/5030469389?check_suite_focus=true.
> 
> It makes me wonder if the changes to the supports won't cause block invalidation. It needs to be tested. It might be not that bad at this stage, but we might need a temporary deprecation if any theme in the directory is using those blocks.

> I could not find any themes in the directory that uses the new set of comments blocks.
I think it would be helpful to have a decision tree to help us know when to do deprecations. And also instructions for how to test them.

I left the background option for not having to deprecate anything 😄

gziolo commented on lib/compat/experimental/blocks.php:
```diff
@@ -106,12 +106,24 @@ function get_comments_pagination_arrow( $block, $pagination_type = 'next' ) {
+			'pageComments'        => get_option( 'page_comments' ),
+			'threadComments'      => get_option( 'thread_comments' ),
+			'threadCommentsDepth' => get_option( 'thread_comments_depth' ),
+			'avatarURL'           => get_avatar_url(
```
I checked the Discussion Settings page and it looks like it has a ton of setting including the default avatar. That was surprising 😄

cbravobernal replied:
Yep, most surprising is that the default avatar of discussion settings is the same default avatar selected for post authors.

gziolo commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -201,11 +207,46 @@ export default function CommentTemplateEdit( {
+		}
+	}
+
 	if ( ! rawComments ) {
```
I think we could move this check just after the `useSelect` call that returns `rawComments` to avoid unnecessary computations. I also see `useMemo` hook call, so it would need to be after that.

gziolo commented on packages/block-library/src/comment-template/edit.js:
```diff
@@ -201,11 +207,46 @@ export default function CommentTemplateEdit( {
 	// Then, we show only a maximum of `perPage` number of comments.
 	// This is because passing `per_page` to `getEntityRecords()` does not
 	// take into account nested comments.
-	const comments = useMemo(
+	let comments = useMemo(
 		() => convertToTree( rawComments ).slice( 0, perPage ),
 		[ rawComments, perPage ]
 	);
 
+	// In case that `threadCommentsDepth` is falsy, we default to a somewhat
+	// arbitrary value of 3.
+	// In case that the value is set but larger than 3 we truncate it to 3.
+	const commentsDepth = Math.min( threadCommentsDepth || 3, 3 );
+
+	// We set a limit in order not to overload the editor of empty comments.
+	const defaultCommentsToShow =
+		perPage <= commentsDepth ? perPage : commentsDepth;
```
This can be moved to the scope of `if` section as it's used only when `postId` is falsy. I commented about that earlier, all the logic for generating the tree of comment placeholders could be in their own function so it documents better.

gziolo reviewed (approved):
I think we are good. I left some non-blocking comments to consider. This enhancement is really exciting. Thank you for finding such a creative and simple solution 👍🏻

Commit:
Moved empty comment creation to function and tried to optimize small code
